man/man1/dust-compiler.1
.TH "DUST\-COMPILER" "1" "October 2013" "" ""
.SH "NAME"
\fBdust-compiler\fR \- Dust\.js template watcher/compiler
.SH SYNOPSIS
.P
dust\-compiler \-s|\-\-source source_path \-d|\-\-destination destination_path
[\-\-bootstrap] [\-\-nonotify] [\-\-includepath] [\-\-amd]
.SH DESCRIPTION
.P
\fBdust\-compiler\fR is a command line tool that will watch and compile Dust\.js
templates\.  It provides notification support for OS X, Linux, and Windows\.
.P
The following options are available:
.P
\-s, \-\-source         Source path for uncompiled Dust templates\.
.P
\-d, \-\-destination    Destination path to write out compiled Dust templates\.
.P
\-\-bootstrap          Process all files immediatley\.
.P
\-\-nonotify           Do not send notification messages\.
.P
\-\-includepath        Include the path of the template as part of the registered
                     name in the compiled template\.
.P
\-\-amd                Wrap the output AMD style
.SH SEE ALSO
.P
https://github\.com/jamsyoung/dust\-compiler
.P
https://npmjs\.org/package/dust\-compiler
.SH AUTHOR
.P
James Young jmeyoung@gmail\.com http://jamsyoung\.com



ChangeLog.md
# Change Log

## 0.3.0
- adding --amd flag for AMD support


## 0.2.0
- restructured to have an internal library (`lib/compile.js`) that could be
  used indpendent of the command line wrapper.

- pulled the log notification pieces out into a stand alone
  [log-notify][log-notify] npm package which is included as a dependency

- added lots of unit tests and mock data to test against

- got blanket coverage working (kinda) - fixes issue #16

- issue #15 should not cause any problems, as far as I can tell.


## 0.1.2
- Updated README


## 0.1.1
- Updated `dustjs-linkedin` to 2.0.3


## 0.1.0
- no functional changes, minor version bumped because it should have always been
  0.1.x from the beginning.

- updated `colors` to 0.6.2

- updated `dustjs-linkedin` to 2.0.2

- added lots of dev dependencies to support grunt test tasks

- added a man page, now you can do `man dust-compiler` from a shell if desired


## 0.0.9
- EXPERIMENTAL - added support for Windows notification via growl ([ZoogieZork][1])

- switched Linux notification to hit notify-send via grow ([ZoogieZork][1])

- upgraded dustjs-linkedin to 1.2.4

- added `-v|--version` argument to display the version installed

- changed how `-s|--source` and `-d|--destination` arguments processed

- reverted `--bootstrap` to only build, it does not watch anymore



## 0.0.8
- code refactoring - made variable names make more sense

- fixed a defect with error messages not being properly displayed by
  terminal-notify in the OS X notification center

- Issue #10 - clearer identification of error messages



## 0.0.7
- modified to not include the .dust extension in the registered name of the
  compiled dust template ([CRHain88][0])



## 0.0.6
- updated `watch` to 0.7.0

- updated `dustjs-linkedin` to 1.2.3

- updated `mkdirp` to 0.3.5

- added `optimist` support for handling command line arguments

- added `colors` support for fancy command line colors when output is written
  to the console

- Issue #4 - added support for recursing directories

- Issue #2 - added `-s|--source` argument for passing in the source directory

- Issue #2 - added `-d|--destination` argument for passing in the destination
  directory

- Issue #8 - added `--includepath` argument to include the path in the compiled
  template as the name

- changed `--no-notify` to `--nonotify`

- improved output to be more verbose

- modified `--bootstrap` to watch after it bootstraps, so no need to re-run the
  compiler after a bootstrap

- Issue #3 - Now traps when source or destination path does not exist



## 0.0.5
- added support notifications on Linux via notify-send

- added `--no-notify` option for platforms that do not have notify-send or
  terminal-notifier support, or people that just don't want to see any
  notifications

- `--bootstrap` option skips directories (vs. crashing)



## 0.0.4
- initial published version



[0]: https://github.com/CRHain88
[1]: https://github.com/ZoogieZork
[log-notify]: https://npmjs.org/package/log-notify



lib/dust-compiler.js
'use strict';

var fs = require('fs'),
    log = require('log-notify'),
    argv,
    path = require('path'),
    watch = require('watch'),
    wrench = require('wrench'),
    compile = require('./compile'),
    metadata = require('../package.json'),
    optimist = require('optimist'),
    sourcePath,
    destinationPath;

require('colors');

argv = optimist
    .usage('Usage: dust-compiler -s|--source source_path -d|--destination destination_path [--bootstrap] [--nonotify] [--includepath]')
    .alias('source', 's')
    .alias('destination', 'd')
    .alias('version', 'v')
    .alias('help', 'h')
    .describe('source', 'The source .dust templates to watch.')
    .describe('destination', 'The destination destination to write compiles .js files out to.')
    .describe('bootstrap', 'Compiles all files on start.')
    .describe('nonotify', 'Do not notify.')
    .describe('includepath', 'Include the path in the name of the compiled dust template.')
    .describe('amd', 'Wrap the output AMD style')
    .describe('version', 'Show the current version installed of dust-compiler')
    .describe('help', 'You are looking at it')
    .argv;

if (argv.v) {
    console.log('dust-compiler v' + metadata.version);
    process.exit(0);
}

if (argv.h) {
    console.log(optimist.help());
    process.exit(0);
}

if (argv.s && argv.d) {
    sourcePath = argv.s.match('/$') ? path.normalize(argv.s) : path.normalize(argv.s + '/');      // must end in slash
    destinationPath = argv.d.match('/$') ? path.normalize(argv.d) : path.normalize(argv.d + '/'); // must end in slash
} else {
    optimist.showHelp();
    console.log('ERROR: '.red + '-s and -d are required arguments');
    process.exit(1);
}

if (argv.nonotify) {
    log = console.log;
}

if (!fs.existsSync(sourcePath)) {
    console.log('ERROR: '.red + sourcePath + ' does not exist');
    process.exit(1);
}

if (!fs.existsSync(destinationPath)) {
    console.log('ERROR: '.red + destinationPath + ' does not exist');
    process.exit(1);
}

if (argv.bootstrap) {
    wrench.readdirRecursive(sourcePath, function (error, fileList) {
        if (error) { console.log(error); }

        if (fileList) {
            fileList.forEach(function (filename) {
                compile(filename, sourcePath, destinationPath, argv);
            });
        }
    });
} else {
    watch.createMonitor(sourcePath, function (monitor) {
        log('Watching '.green + sourcePath);
        log('Will write to '.yellow + destinationPath.yellow);
        log('Type ^C to cancel'.yellow);

        monitor.on('created', function (filename) {
            compile(filename, sourcePath, destinationPath, argv);
        });

        monitor.on('changed', function (filename) {
            compile(filename, sourcePath, destinationPath, argv);
        });
    });
}
